src/main/java/net/fabricmc/loom/processors/MinecraftProcessedProvider.java

```diff
@@ -50,23 +57,54 @@ public MinecraftProcessedProvider(Project project, JarProcessorManager jarProces
 
 	@Override
 	protected void addDependencies(DependencyInfo dependency, Consumer<Runnable> postPopulationScheduler) {
-		if (jarProcessorManager.isInvalid(projectMappedJar) || isRefreshDeps()) {
+		boolean invalid;
+
+		if (this.split) {
+			invalid = this.jarProcessorManager.isInvalid(Environment.COMPILE, this.projectMappedCompileJar) || this.jarProcessorManager.isInvalid(Environment.RUNTIME, this.projectMappedRuntimeJar);
+		} else {
+			invalid = this.jarProcessorManager.isInvalid(Environment.BOTH, this.projectMappedCommonJar);
+		}
+
+		if (invalid || isRefreshDeps()) {
 			getProject().getLogger().lifecycle(":processing mapped jar");
 			invalidateJars();
 
 			try {
-				FileUtils.copyFile(super.getMappedJar(), projectMappedJar);
+				File mappedJar = super.getMappedJar();
+				File firstProcessedJar = this.getMappedJar();
+
+				FileUtils.copyFile(mappedJar, firstProcessedJar);
+
+				this.jarProcessorManager.process(Environment.BOTH, firstProcessedJar);
+
+				if (this.split) {
+					FileUtils.copyFile(firstProcessedJar, this.projectMappedRuntimeJar);
+
+					this.jarProcessorManager.process(Environment.COMPILE, this.projectMappedCompileJar);
+					this.jarProcessorManager.process(Environment.RUNTIME, this.projectMappedRuntimeJar);
+				}
 			} catch (IOException e) {
-				throw new RuntimeException("Failed to copy source jar", e);
-			}
+				String message = "Failed to copy or process source jar";
 
-			jarProcessorManager.process(projectMappedJar);
+				if (this.split) {
+					message += "s";
+				}
+
+				throw new UncheckedIOException(message, e);
+			}
 		}
 
 		getProject().getRepositories().flatDir(repository -> repository.dir(getJarDirectory(getExtension().getProjectPersistentCache(), PROJECT_MAPPED_CLASSIFIER)));
 
-		getProject().getDependencies().add(Constants.Configurations.MINECRAFT_NAMED,
-				getProject().getDependencies().module("net.minecraft:minecraft:" + getJarVersionString(PROJECT_MAPPED_CLASSIFIER)));
+		String compileClassifier = this.split ? PROJECT_MAPPED_COMPILE_CLASSIFIER : PROJECT_MAPPED_CLASSIFIER;
+		String runtimeClassifier = this.split ? PROJECT_MAPPED_RUNTIME_CLASSIFIER : PROJECT_MAPPED_CLASSIFIER;
+
+		getProject().getDependencies().add(Constants.Configurations.MINECRAFT_NAMED_COMPILE,
+				getProject().getDependencies().module("net.minecraft:minecraft:" + getJarVersionString(compileClassifier))
+		);
+		getProject().getDependencies().add(Constants.Configurations.MINECRAFT_NAMED_RUNTIME,
+				getProject().getDependencies().module("net.minecraft:minecraft:" + getJarVersionString(runtimeClassifier))
```

Comment:
As I said in the other PR it should only use the split jars when required.

Reply:
That is what it does. If there is not a need to split the JARs, then `compileClassifier == runtimeClassifier` and this adds the single JAR to compilation and runtime class paths.

Reply:
Ah ok, I see. as long as IDE's dont show it as 2 entries (duplicated search for example) then this should be fine.

Reply:
That is a problem. I'll have to hide the runtime JAR somehow. :thinking: